### PR TITLE
[dualtor]: Add normal operation test cases

### LINKED ISSUE #2760: Normal Operation Test Case
Test normal operation for all three packet flow scenarios

Depends on #2758, #2759

### LINKED ISSUE #2761: Config Reload Test Case
Test all packet flow scenarios after a config reload for each ToR (see test plan action items doc for more detail)

Depends on #2758, #2759, #2755

### LINKED ISSUE #2762: Manual ToR Switch Test Case
Test all packet flow scenarios after manually switching the active ToR

Depends on #2758, #2759, #2755

### PR DESCRIPTION
**Pull request by theasianpianist**
Signed-off-by: Lawrence Lee <lawlee@microsoft.com>

### Description of PR

Summary:
Closes #2760
Closes #2761
Closes #2762 

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### How did you do it?
* Covers normal operation, config reload, and CLI tor switchover

Send traffic and perform each of the above scenarios, for all three packet flows (upstream, downstream via active, downstream via standby). Uses `data_plane_utils.py` to verify that traffic disruption was within limits, and checks APP and STATE DB on both ToRs to verify correct state after the test.

#### How did you verify/test it?
```
=== Running tests in groups ===
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
=================================================================================== test session starts ====================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.10.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /var/src/sonic-mgmt/tests, inifile: pytest.ini
plugins: forked-1.3.0, metadata-1.11.0, xdist-1.28.0, html-1.22.1, repeat-0.9.1, ansible-2.2.2
collected 10 items

dualtor/test_normal_op.py::test_normal_op_upstream PASSED                                                                                                                            [ 10%]
dualtor/test_normal_op.py::test_normal_op_downstream_active PASSED                                                                                                                   [ 20%]
dualtor/test_normal_op.py::test_normal_op_downstream_standby PASSED                                                                                                                  [ 30%]
dualtor/test_normal_op.py::test_active_config_reload_upstream PASSED                                                                                                                 [ 40%]
dualtor/test_normal_op.py::test_standby_config_reload_upstream PASSED                                                                                                                [ 50%]
dualtor/test_normal_op.py::test_standby_config_reload_downstream_active PASSED                                                                                                       [ 60%]
dualtor/test_normal_op.py::test_active_config_reload_downstream_standby
-------------------------------------------------------------------------------------- live log call ---------------------------------------------------------------------------------------
22:23:49 ERROR dual_tor_io.py:start_io_test:171: The longest disruption lasted 0.033 seconds.11 packet(s) lost.
22:23:49 ERROR dual_tor_io.py:start_io_test:174: Total disruptions count is 269. All disruptions lasted 2.927 seconds. Total 537 packet(s) lost
FAILED                                                                                                                                                                               [ 70%]
dualtor/test_normal_op.py::test_tor_switch_upstream PASSED                                                                                                                           [ 80%]
dualtor/test_normal_op.py::test_tor_switch_downstream_active
-------------------------------------------------------------------------------------- live log call ---------------------------------------------------------------------------------------
22:41:51 ERROR dual_tor_io.py:start_io_test:171: The longest disruption lasted 0.020 seconds.2 packet(s) lost.
22:41:51 ERROR dual_tor_io.py:start_io_test:174: Total disruptions count is 298. All disruptions lasted 1.224 seconds. Total 308 packet(s) lost
FAILED                                                                                                                                                                               [ 90%]
dualtor/test_normal_op.py::test_tor_switch_downstream_standby
-------------------------------------------------------------------------------------- live log call ---------------------------------------------------------------------------------------
22:51:25 ERROR dual_tor_io.py:start_io_test:171: The longest disruption lasted 0.041 seconds.11 packet(s) lost.
22:51:25 ERROR dual_tor_io.py:start_io_test:174: Total disruptions count is 284. All disruptions lasted 2.285 seconds. Total 580 packet(s) lost
FAILED                                                                                                                                                                               [100%]
```
The large disruption count is due to the fact that current dataplane testing code does not measure traffic disruption per-server, but does it for all servers as a whole. Since the packets are randomly sent to different servers, if non-consecutive packets sent to a single server that is switching over are dropped, the test reports each packet drop as a separate disruption rather than combining them. This behavior will be fixed in an upcoming PR, but for now these test results are not cause for alarm.

### CODE REVIEW

File: tests/dualtor/test_normal_op.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    expect_db_values(upper_tor_host, APP_DB, 'active')
+    expect_db_values(upper_tor_host, STATE_DB, 'active', 'healthy')
+    expect_db_values(lower_tor_host, APP_DB, 'standby')
+    expect_db_values(lower_tor_host, STATE_DB, 'standby', 'healthy')
+
+
+def test_normal_op_downstream_standby(upper_tor_host, lower_tor_host, send_t1_to_server_with_action, toggle_all_simulator_ports_to_upper_tor):
+    """Send downstream traffic to the standby ToR and confirm no disruption or switchover occurs"""
+    send_t1_to_server_with_action(lower_tor_host, verify=True)
+    expect_db_values(upper_tor_host, APP_DB, 'active')
```
Comment (vaibhavhd) [inline tests/dualtor/test_normal_op.py:35]: Cosmetic change suggestion - a helper function can be used if you think it improves readability/simplicity.

```Python
def verify_db_values(activehost, standbyhost):
    expect_db_values(activehost, APP_DB, 'active')
    expect_db_values(activehost, STATE_DB, 'active', 'healthy')
    expect_db_values(standbyhost, APP_DB, 'standby')
    expect_db_values(standbyhost, STATE_DB, 'standby', 'healthy')
```
```Python
def test_normal_op_downstream_standby(upper_tor_host, lower_tor_host, send_t1_to_server_with_action, toggle_all_simulator_ports_to_upper_tor):
    """Send downstream traffic to the standby ToR and confirm no disruption or switchover occurs"""
    send_t1_to_server_with_action(lower_tor_host, verify=True)
    # expect no switchover
    verify_db_values(activehost=upper_tor_host, standbyhost=lower_tor_host)


def test_tor_switch_upstream(upper_tor_host, lower_tor_host, send_server_to_t1_with_action, toggle_all_simulator_ports_to_upper_tor, force_active_tor):
    """Send upstream traffic and perform switchover via CLI. Confirm switchover occurs and disruption lasts < 1 second"""
    send_server_to_t1_with_action(upper_tor_host, standbyhost=lower_tor_host, verify=True, delay=1, action=lambda: force_active_tor(lower_tor_host, 'all'))
    # expect switchover
    verify_db_values(activehost=lower_tor_host, standbyhost=upper_tor_host)
```

File: tests/dualtor/test_normal_op.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+from tests.common.dualtor.mux_simulator_control import toggle_all_simulator_ports_to_upper_tor, toggle_all_simulator_ports_to_lower_tor         # lgtm[py/unused-import] 
+from tests.common.fixtures.ptfhost_utils import run_icmp_responder, run_garp_service, copy_ptftests_directory, change_mac_addresses             # lgtm[py/unused-import]
+
+pytestmark = [
+    pytest.mark.topology("dualtor")
+]
+
+def test_normal_op_upstream(upper_tor_host, lower_tor_host, send_server_to_t1_with_action, toggle_all_simulator_ports_to_upper_tor):
+    """Send upstream traffic and confirm no disruption or switchover occurs"""
+    send_server_to_t1_with_action(upper_tor_host, standbyhost=lower_tor_host, verify=True)
```
[the remark this answers is not in the document]
Comment (theasianpianist) [reply tests/dualtor/test_normal_op.py:16]: ack

File: tests/dualtor/test_normal_op.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    """Send downstream traffic to the active ToR and `config reload` the standby ToR. Confirm no switchover occurs and no disruption"""
+    send_t1_to_server_with_action(upper_tor_host, standbyhost=lower_tor_host, verify=True, action=lambda: config_reload(lower_tor_host, wait=0))
+    expect_db_values(upper_tor_host, APP_DB, 'active')
+    expect_db_values(upper_tor_host, STATE_DB, 'active', 'healthy')
+    expect_db_values(lower_tor_host, APP_DB, 'standby')
+    expect_db_values(lower_tor_host, STATE_DB, 'standby', 'healthy')
+
+
+def test_active_config_reload_downstream_standby(upper_tor_host, lower_tor_host, send_t1_to_server_with_action, toggle_all_simulator_ports_to_upper_tor):
+    """Send downstream traffic to the standby ToR and `config reload` the active ToR. Confirm no switchover occurs and no disruption"""
```
[the remark this answers is not in the document]
Comment (theasianpianist) [reply tests/dualtor/test_normal_op.py:69]: Good catch, copied the wrong docstring!

File: tests/dualtor/test_normal_op.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+def test_normal_op_upstream(upper_tor_host, lower_tor_host, send_server_to_t1_with_action, toggle_all_simulator_ports_to_upper_tor):
+    """Send upstream traffic and confirm no disruption or switchover occurs"""
+    send_server_to_t1_with_action(upper_tor_host, standbyhost=lower_tor_host, verify=True)
+    expect_db_values(upper_tor_host, APP_DB, 'active')
+    expect_db_values(upper_tor_host, STATE_DB, 'active', 'healthy')
+    expect_db_values(lower_tor_host, APP_DB, 'standby')
+    expect_db_values(lower_tor_host, STATE_DB, 'standby', 'healthy')
+
+
+def test_normal_op_downstream_active(upper_tor_host, lower_tor_host, send_t1_to_server_with_action, toggle_all_simulator_ports_to_upper_tor):
```
[the remark this answers is not in the document]
Comment (theasianpianist) [reply tests/dualtor/test_normal_op.py:23]: Ack

File: tests/dualtor/test_normal_op.py (hunk 1 above)
Comment (theasianpianist) [reply tests/dualtor/test_normal_op.py:35]: Ack, will create in `control_plane_utils.py`. Good idea!

File: tests/dualtor/test_normal_op.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
+import pytest 
+
+from tests.common.config_reload import config_reload
+from tests.common.dualtor.control_plane_utils import expect_db_values, APP_DB, STATE_DB
+from tests.common.dualtor.data_plane_utils import send_t1_to_server_with_action, send_server_to_t1_with_action                                  # lgtm[py/unused-import]
+from tests.common.dualtor.dual_tor_utils import *                                                                                               # lgtm[py/unused-import]
```
[the remark this answers is not in the document]
Comment (theasianpianist) [reply tests/dualtor/test_normal_op.py:6]: ack

File: tests/common/dualtor/control_plane_utils.py [hunk 6]
```diff
@@ -85,5 +96,27 @@ def expect_db_values(duthost, db, state, health=None, intf_names='all'):
...
+    """
+
+    expect_db_values(expected_active_host, APP_DB,
+                     'active', intf_names=intf_names)
+    expect_db_values(expected_active_host, STATE_DB, 'active',
+                     'healthy', intf_names=intf_names)
+    expect_db_values(expected_standby_host, APP_DB,
+                     'standby', intf_names=intf_names)
+    expect_db_values(expected_standby_host, STATE_DB, 'standby',
+                     expected_standby_health, intf_names=intf_names)
```
Comment (vaibhavhd) [inline tests/common/dualtor/control_plane_utils.py:122]: Was the line at EOF removed?

**Review by vaibhavhd — APPROVED** [review]
Thanks for PIP8 compliance changes.

**Comment by vaibhavhd** [comment]
Can we add some results in the `How did you verify/test it?` section. It would be interesting if we can capture the output of any test which might be failing at the moment?

File: tests/common/dualtor/control_plane_utils.py (hunk 6 above)
Comment (theasianpianist) [reply tests/common/dualtor/control_plane_utils.py:122]: There is a blank line, but Github doesn't display it. It displays the ⛔ symbol if it's not present I think

**Pull request merged**